capa/features/extractors/ghidra/file.py

```diff
@@ -0,0 +1,198 @@
+
+import struct
+from typing import Tuple, Iterator
+
+from ghidra.program.model.symbol import SourceType
+
+import capa.features.extractors.common
+import capa.features.extractors.helpers
+import capa.features.extractors.strings
+import capa.features.extractors.ghidra.helpers
+from capa.features.file import Export, Import, Section, FunctionName
+from capa.features.common import FORMAT_PE, FORMAT_ELF, Format, String, Feature, Characteristic
+from capa.features.address import NO_ADDRESS, Address, FileOffsetAddress, AbsoluteVirtualAddress
+
+MAX_OFFSET_PE_AFTER_MZ = 0x200
+
+
+def check_segment_for_pe(mem_block) -> Iterator[Tuple[int, int]]:
+    """check segment for embedded PE
+
+    adapted for Ghidra from:
+    https://github.com/vivisect/vivisect/blob/91e8419a861f4977https://github.com/vivisect/vivisect/blob/91e8419a861f49779f18316f155311967e696836/PE/carve.py#L259f18316f155311967e696836/PE/carve.py#L25
+    """
+
+    mz_xor = [
+        (
+            capa.features.extractors.helpers.xor_static(b"MZ", i),
+            capa.features.extractors.helpers.xor_static(b"PE", i),
+            i,
+        )
+        for i in range(256)
+    ]
+ 
+    todo = []
+    for mzx, pex, i in mz_xor:
+        # find all segment offsets containing XOR'd "MZ" bytes
+        for off in capa.features.extractors.ghidra.helpers.find_byte_sequence(mzx):
+            todo.append((off, mzx, pex, i))
+
+    seg_max = mem_block.getEnd() 
+    while len(todo):
+        off, mzx, pex, i = todo.pop()
+
+        # MZ header has one field we will check e_lfanew is at 0x3c
+        e_lfanew = off.add(0x3C)
+
+        if seg_max.getOffset() < (e_lfanew.getOffset() + 4):
+            continue
+
+        e_lfanew_bytes = b''
+        e_lfanew_sbytes = getBytes(e_lfanew, 4) 
+        for b in e_lfanew_sbytes:
+            b = (b & 0xFF).to_bytes(1, 'little')
+            e_lfanew_bytes = e_lfanew_bytes + b
+        newoff = struct.unpack("<I", capa.features.extractors.helpers.xor_static(e_lfanew_bytes, i))[0]
+
+        # assume XOR'd "PE" bytes exist within threshold
+        if newoff > MAX_OFFSET_PE_AFTER_MZ:
+            continue
+
+        peoff = off.add(newoff)
+        if seg_max.getOffset() < (peoff.getOffset() + 2):
+            continue
+
+        if getBytes(peoff, 2) == pex:
+            yield off, i
+
+
+def extract_file_embedded_pe() -> Iterator[Tuple[Feature, Address]]:
+    """extract embedded PE features
+    """
+    for seg in currentProgram.getMemory().getBlocks():
+        for ea, _ in check_segment_for_pe(seg):
+            yield Characteristic("embedded pe"), FileOffsetAddress(ea)
+
+
+def extract_file_export_names() -> Iterator[Tuple[Feature, Address]]:
+    """extract function exports
+
+    Exports in Ghidra:
+        - namespace = Global
+        - SourceType = IMPORTED
+    """
+
+    fm = currentProgram.getFunctionManager().getFunctions(True) # Only Global functions
+    while fm.hasNext():
+        f = fm.next().getSymbol()
```

Comment:
We could possibly skip some indirection here by using `currentProgram.getSymbolTable().getExternalEntryPointIterator()` and `currentProgram.getSymbolTable().getPrimarySymbol(_addr_)`

https://ghidra.re/ghidra_docs/api/ghidra/program/model/symbol/SymbolTable.html#getExternalEntryPointIterator()
https://ghidra.re/ghidra_docs/api/ghidra/program/model/symbol/SymbolTable.html#getPrimarySymbol(ghidra.program.model.address.Address)